diff --git a/clutter/clutter/clutter-frame-clock.c b/clutter/clutter/clutter-frame-clock.c
index 16098b70f..95e30d2f4 100644
--- a/clutter/clutter/clutter-frame-clock.c
+++ b/clutter/clutter/clutter-frame-clock.c
@@ -37,7 +37,7 @@ static guint signals[N_SIGNALS];
 /* An estimate queue holds several int64_t values. Adding a new value to the
  * queue overwrites the oldest value.
  */
-#define ESTIMATE_QUEUE_LENGTH 16
+#define ESTIMATE_QUEUE_LENGTH 256
 
 typedef struct _EstimateQueue
 {
